Fix #10386: Fix `created_on` in `UserSettingsModel`

sajalasati opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes #10386.
2. This PR does the following:

Fixes the `created_on` in `UserSettingsModel` via one-off job. The job looks through other fields in `UserSettingModel` and through other models from the same user (in the `core.storage.user` module) and compares them in order to find the lower date by checking their date-time attributes if they have any. This date is then set as the new `created_on`. If the new date is within **5 minutes** from the current value of `created_on`, it is not updated, else it will be updated.

Following models and attributes have been considered based on the test runs of the job on the backup server (Find my analysis [here](https://docs.google.com/spreadsheets/d/1rkrarSpwKbPMetMEe6YFmLU1Dd7HK-Qqr8XrO0e4V_w/edit?usp=sharing) of all the runs):
- `ExplorationUserDataModel`: rated_on, last_updated, created_on
- `UserContributionsModel`, `UserEmailPreferencesModel`, `UserStatsModel`, `UserSubscriptionsModel`: only last_updated, created_on
- `UserSettingsModel`: All datetime attributes including first_contribution_msec (see detail in code)

## Essential Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#instructions-for-making-a-code-change).
- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.

Commit:
update created_on attribute code with tests

Commit:
Merge remote-tracking branch 'upstream/develop' into correct-created-on-attribute

Commit:
add one-off job to registry file

sajalasati commented:
Hi @seanlip, PTAL! Also tell whether I should fill in the [server jobs form](https://goo.gl/forms/XIj00RJ2h5L55XzU2) now or wait for at least one approving review?

seanlip commented:
Hi @sajalasati, I think let's get the review from @vojtechjelinek, and he will determine if it makes sense to fill out the form. Thanks!

vojtechjelinek commented on core/domain/user_jobs_one_off.py:
```diff
@@ -468,6 +468,148 @@ def reduce(key, values):
+
+    @staticmethod
+    def map(user_settings_model):
+
```
Remove this empty line.

sajalasati replied:
Done.

vojtechjelinek commented on core/domain/user_jobs_one_off.py:
```diff
@@ -468,6 +468,148 @@ def reduce(key, values):
 
+class FixUserSettingsCreatedOnOneOffJob(jobs.BaseMapReduceOneOffJobManager):
+    """Job that fixes the invalid values of created_on attribute in the
+    UserSettingsModel.
```
Please explain that this job is just one off and can be removed from the codebase after we resolbe the issue with `created_on`.

sajalasati replied:
Thanks. I have added that statement now. Does it look okay?

vojtechjelinek commented on core/domain/user_jobs_one_off.py:
```diff
@@ -468,6 +468,148 @@ def reduce(key, values):
+    UserSettingsModel.
+    """
+
+    @classmethod
```
Add this above
```
    def enqueue(cls, job_id, additional_job_params=None):
        super(FixUserSettingsCreatedOnOneOffJob, cls).enqueue(
			job_id, shard_count=64)
```

sajalasati replied:
I have added it now.

Though I also want to understand in general when we should override the `enqueue` method, i.e when should I increase the `shard_count`. I only understand that it's used to process things in parallel. The test classes using it say that "we can raise the number of shards for this job, since it goes only over one type of entity class", but I found other test classes also going through one entity class type but not using it. Couldn't deduce anything solid from what I observed. Just curious to know. Thanks :)

vojtechjelinek replied:
Yep, we implemented this "quite" recently (around 9 months) and only add it to jobs that were created from that time and where we remembered about this.

sajalasati replied:
Ok, thanks for the info!

vojtechjelinek commented on core/domain/user_jobs_one_off.py:
```diff
@@ -468,6 +468,148 @@ def reduce(key, values):
+            user_models.ExpUserLastPlaythroughModel.query(
+                user_models.ExpUserLastPlaythroughModel.user_id == user_id)
+            .get()
+        )
```
```suggestion
        exp_user_last_playthrough_model = (
            user_models.ExpUserLastPlaythroughModel.query(
                user_models.ExpUserLastPlaythroughModel.user_id == user_id
            ).get()
        )
```

sajalasati replied:
Done.

vojtechjelinek commented on core/domain/user_jobs_one_off.py:
```diff
@@ -468,6 +468,148 @@ def reduce(key, values):
+            user_models.UserContributionProficiencyModel.query(
+                user_models.UserContributionProficiencyModel.user_id == user_id)
+            .get()
+        )
```
```suggestion
        user_contribution_proficiency_model = (
            user_models.UserContributionProficiencyModel.query(
                user_models.UserContributionProficiencyModel.user_id == user_id
            ).get()
        )
```

sajalasati replied:
Done.

vojtechjelinek commented on core/domain/user_jobs_one_off.py:
```diff
@@ -468,6 +468,148 @@ def reduce(key, values):
+            user_settings_model.created_on = min_date
+            user_settings_model.update_timestamps()
+            user_settings_model.put()
+            yield ('SUCCESS_UPDATED', 1)
```
Can we actually log here from what model we extracted the value? So that we can than see if we actually need to analyze all the models or just a subset.

sajalasati replied:
I didn't get you. Are we going to analyse other models? Because so far I understood that the problem lied with `UserSettingsModel`, that's why it doesn't have the correct timestamps. Other models I am looking at could have been created at any time, not necessarily at the start. What you are asking is doable, but I am only trying to understand the motive behind it. Thanks!

vojtechjelinek replied:
Basically, I expect that when we run this on the backup server we will find that the lowest value is coming only from a handful of models if that is the truth we can then decided to leave some models out when we run this on the production in order to make this job more efficient.

sajalasati replied:
I have made the change now. PTAL!

vojtechjelinek commented on core/domain/user_jobs_one_off.py:
```diff
@@ -468,6 +468,148 @@ def reduce(key, values):
+        min_date = min(filtered_user_dates_list)
+        if min_date < user_settings_model.created_on:
+            user_settings_model.created_on = min_date
+            user_settings_model.update_timestamps()
```
```suggestion
            user_settings_model.update_timestamps(update_last_updated_time=False)
```

sajalasati replied:
Done.

vojtechjelinek commented on core/domain/user_jobs_one_off_test.py:
```diff
@@ -1585,6 +1585,272 @@ def test_one_setting_model_without_gae_id(self):
+            user_settings_model.created_on, final_created_on_timestamp)
+
+        output = self._run_one_off_job()
+        # There already exists a UserSettings created in test_utils. It's
+        # corresponding user_auth_details_model is created first (which has
+        # a lower timestamp), hence it is also being updated here.
```
You can use this trick to get rid of these comments https://github.com/oppia/oppia/blob/6bbfd2240602e4b3af40d361900337b36dab6227/core/domain/user_jobs_one_off_test.py#L2170-L2179

sajalasati replied:
This makes it quite easy to test the code. Thanks for suggesting this! I have changed it now.

vojtechjelinek commented on core/domain/user_jobs_one_off_test.py:
```diff
@@ -1585,6 +1585,272 @@ def test_one_setting_model_without_gae_id(self):
+        user_settings_model.last_created_an_exploration = (
+            original_created_on_timestamp + datetime.timedelta(hours=2))
+
+        final_created_on_timestamp = user_settings_model.last_updated
```
Could you add comment above this explaining the we did not modify `last_updated` and because of that it is now the minimal value. It took me around a minute to figure out why exactly `last_updated`.

sajalasati replied:
Sure, I have added a comment saying that now. Added one or two comments elsewhere too where I felt similar.

vojtechjelinek commented on core/domain/user_jobs_one_off_test.py:
```diff
@@ -1585,6 +1585,272 @@ def test_one_setting_model_without_gae_id(self):
+        user_settings_model_1.update_timestamps()
+        original_created_on_timestamp_1 = user_settings_model_1.created_on
+        user_settings_model_1.created_on += datetime.timedelta(hours=10)
+        final_created_on_timestamp_1 = user_settings_model_1.last_updated
```
Should we test other field of `UserSettingsModel` for the lowest value?

sajalasati replied:
You mean some other date-time attribute of `UserSettingsModel`? I printed out the value of all those attributes in this test, and found that only `created_on` and `last_updated` exist, and rest all are `None`. So that's why here only `last_updated` attribute could be the answer. Is this what you wanted me to verify?

vojtechjelinek replied:
I mean to manualy set some other datatime field of `UserSettingsModel` to the lowest value to check that the extraction works from other fields works too.

sajalasati replied:
Okay, so I have created a separate test for that now - one test using `last_updated` and another using other attributes of `UserSettingsModel` for updating correct timestamp.

vojtechjelinek commented on core/domain/user_jobs_one_off_test.py:
```diff
@@ -1585,6 +1585,272 @@ def test_one_setting_model_without_gae_id(self):
+
+        migrated_user_model = (
+            user_models.UserSettingsModel.get_by_id(self.USER_ID_1))
+
```
Remvoe this empty line.

sajalasati replied:
Done.

vojtechjelinek commented on core/domain/user_jobs_one_off_test.py:
```diff
@@ -1585,6 +1585,272 @@ def test_one_setting_model_without_gae_id(self):
+        user_settings_model.last_updated += datetime.timedelta(hours=10)
+        user_settings_model.put()
+
+        final_created_on_timestamp = (
+            original_created_on_timestamp + datetime.timedelta(hours=1))
```
I think doing `final_created_on_timestamp = user_subscriptions_model.last_checked` down below and instead of `last_checked=final_created_on_timestamp` do `last_checked=original_created_on_timestamp + datetime.timedelta(hours=1)`

sajalasati replied:
It still functions in the same way. I did it my way so that I define what the `final_created_on_timestamp` would be in the starting, and then assign all the other date-time parameters according to it. If you are suggesting this to improve the readability, I will do the change. But I want to know if there's any other reason apart from that. Thanks!
**Edit**: Did the change you suggested

vojtechjelinek reviewed (changes requested):
Thanks! Looks pretty good! One bigger request otherwise just small comments.

Commit:
address comments

sajalasati commented:
@vojtechjelinek PTAL again! I have made changes acc to some of your suggestions, and but I have questions in the rest (I guess including the big one you were referring to about logging the model from which we took min timestamp value). Thanks!

vojtechjelinek commented:
@sajalasati Replied to your comments, I will do a full pass after you make these changes.

Commit:
address follow-up comments

Commit:
Merge remote-tracking branch 'upstream/develop' into correct-created-on-attribute

sajalasati commented:
@vojtechjelinek PTAL now! I have made all the changes you requested for.

vojtechjelinek commented on core/domain/user_jobs_one_off.py:
```diff
@@ -468,6 +468,172 @@ def reduce(key, values):
 
 
+class FixUserSettingsCreatedOnOneOffJob(jobs.BaseMapReduceOneOffJobManager):
+    """Job that fixes the invalid values of created_on attribute in the
+    UserSettingsModel. It is a one-off job and can be removed from the codebase
+    after we resolve this issue by running the job once.
+    """
```
```suggestion
    """Job that fixes the invalid values of created_on attribute in the
    UserSettingsModel. 
    
    This is a one-off job and can be removed from the codebase
    after we resolve this issue by running the job once.
    """
```

sajalasati replied:
Done.

vojtechjelinek commented on core/domain/user_jobs_one_off.py:
```diff
@@ -468,6 +468,172 @@ def reduce(key, values):
+            user_settings_model.first_contribution_msec
+        ]
+
+        model_names = ['UserSettingsModel'] * 9
```
I think it would be better to do something like

```
 user_dates_list = [
    ('UserSettingsModel', user_settings_model.created_on),
    ('UserSettingsModel', user_settings_model.last_updated),
    ('UserSettingsModel', user_settings_model.last_agreed_to_terms),
    …
]
```

and then down below do `model_name, min_date = min(filtered_user_dates_list, key=lambda x: x[1])`, that way you can have just one list with tuples instead of keeping two lists.

sajalasati replied:
Are you suggesting to eliminate `models_linked_with_user_settings_model` by this? I don't get how this will work.
This approach is good for `user_settings_model` since we know that it exists, but the same is not true for other models. I have to check for other models, and only when it exists for the user, I can append its date-time attribute to the list. That is why I thought of keeping models and their names as a list of tuples so that I can loop over it, instead of having to repeatedly write similar code for checking each model (i.e to avoid writing `if model is not None` condition after fetching each model). Thoughts @vojtechjelinek?

If the later approach is what you are suggesting, I will make the change.

vojtechjelinek replied:
What you can do instead of
```
for model_name, model in models_linked_with_user_settings_model:
    if model is not None:
        user_dates_list.append(model.last_updated)
        user_dates_list.append(model.created_on)
        model_names.append(model_name)
        model_names.append(model_name)
```
is something like 
```
for model_name, model in models_linked_with_user_settings_model:
    if model is not None:
        user_dates_list.append((model_name, model.last_updated))
        user_dates_list.append((model_name, model.created_on))
```
right?

vojtechjelinek replied:
So there would be no need for any additional code I guess.

sajalasati replied:
My bad, I didn't get that earlier. Thanks for the explanation. Did the change now.

vojtechjelinek reviewed (changes requested):
Thanks! Few other comments.

Commit:
address nit change

Commit:
address comments

Commit:
Merge remote-tracking branch 'upstream/develop' into correct-created-on-attribute

sajalasati commented:
Hi @vojtechjelinek! I have made all the changes now. PTAL!

sajalasati commented:
Hi @BenHenning! I think now you can have a look at the PR too! I am waiting for an approval from @vojtechjelinek, though I think I have done almost all the changes he had requested for. 

@vojtechjelinek Please also tell if it makes sense to submit the server jobs form once you approve the PR, as @seanlip has suggested for doing that.

Thanks!

vojtechjelinek reviewed (approved):
LGTM! You can submit the form for this job.

BenHenning commented:
Hi! Sorry, will need to take a look at this tomorrow my time.

sajalasati commented:
@seanlip Just FYI, I have submitted the server jobs form.

seanlip commented on core/domain/user_jobs_one_off.py:
```diff
@@ -468,6 +468,184 @@ def reduce(key, values):
+        ]
+
+        # Models in user storage module keyed by user_id.
+        completed_activities_model = (
```
Please use fetch_multiple_entities_by_ids_and_models() in core/platform/datastore for this instead, rather than getting each model one at a time. The latter is very time/resource-consuming, especially given that you're running this repeatedly (once for each user).

sajalasati replied:
Did the change now. Thanks!

seanlip commented on core/domain/user_jobs_one_off.py:
```diff
@@ -468,6 +468,184 @@ def reduce(key, values):
+            user_settings_model.created_on = min_date
+            user_settings_model.update_timestamps(
+                update_last_updated_time=False)
+            user_settings_model.put()
```
For now, how about running this more as an audit job to get info? If you find a discrepancy, output the uid and enough information to debug the discrepancy (e.g. "user xxx created on yyy but field zzz of model www is ttt"). You can even make a list of all the datetimes for all the models if it's useful for analysis. Let's then look at the results before finalizing on what to do (unless this analysis has already been done previously?).

seanlip replied:
Also, if we know there are a lot of these, then you can do something like: generate random number in [0,1], if it's < p (for some chosen value of p) then yield full thing, else just yield an indicator that it was updated.

vojtechjelinek replied:
Yes, we know there is a very high number of `UserSettingsModel` that have the `created_on` field broken. I don't think that this kind of output you mention above will help is with this issue anyhow, but I think that just reporting the data (as the yields are set up now) and skipping actually putting the data, might be a good first run.

vojtechjelinek replied:
The only thing that we can maybe try to confirm that there are no discrepancies after June 2020 (when the issue was fix by #9252).

seanlip replied:
Would it be useful to have the data available so that we can make a hypothesis and verify it? E.g. we might find, from such an analysis, that the source of truth should be model XXX, and then we can do a check to verify that XXX always has the lowest recorded time among all related models, and then use it as the primary source.

vojtechjelinek replied:
Yeah, I see, this is almost what we have implemented, only we do no track the field that was used.

@sajalasati Could you also which field in some model has the lowest value and disable the modification of the data for the first run of the job?

sajalasati replied:
@vojtechjelinek Can you confirm that we just want something like `UserSettingsModel_first_contribution_msec` if this model and field were used for updating, or we want to log all updates like @seanlip suggested intially? (I got a bit confused reading through all the comments, though I think the yield statements we have currently should be sufficient for the analysis)

sajalasati replied:
@vojtechjelinek Also, I think I should keep a time delta value, such that for a user_id if `x` = UserSettingsModel's lowest timestamp, and `y` = the lowest timestamp we found out among all models, then we update the `created_on` only when `x-y >= time_delta` (when y was smaller). 

This will help us avoid a lot of unnecessary updates, like currently in `user_services.create_new_user` ([here](https://github.com/oppia/oppia/blob/develop/core/domain/user_services.py#L1276-L1278)), first the UserAuthDetails and UserIdentifiers models are created, and UserSettingsModel is created after these two, but the current job will still consider it for an update. (I know it happens because in the tests the default created user from test_utils was getting updated).

@vojtechjelinek I want you to help me with deciding what that `time_delta` should be? Something like 5 mins looks good?

vojtechjelinek replied:
> @vojtechjelinek Can you confirm that we just want something like `UserSettingsModel_first_contribution_msec` if this model and field were used for updating, or we want to log all updates like @seanlip suggested intially? (I got a bit confused reading through all the comments, though I think the yield statements we have currently should be sufficient for the analysis)

Yes, something like `UserSettingsModel_first_contribution_msec`, I think this format will be easiest to analyze.

The delta is a nice catch, I think 5 minutes should be fine.

sajalasati replied:
@seanlip @vojtechjelinek I have made the changes now as per [this](https://github.com/oppia/oppia/pull/11437#discussion_r548483548) comment.

sajalasati replied:
> The only thing that we can maybe try to confirm that there are no discrepancies after June 2020 (when the issue was fix by #9252).

@vojtechjelinek Sorry I forgot about this - Do we want to yield error statement for these user_models too i.e some sort of error key for models created after June 2020 which don't have their `created_on` properly set? If yes, then I am planning to yield something like `('ERROR_FOR_USER_MODEL', 'user_id')` for those. Please also tell the exact timestamp I should use for that.

vojtechjelinek replied:
Let's do it from July 1st 2020, but please even if these are broken keep the info about which model has lower value (like with models that are broken before July) so that we do not lose any important info.

sajalasati replied:
(Everywhere by cut-off date I mean July 1st, 2020)

@vojtechjelinek Sorry I did not realise this earlier, but it's **difficult to identify user models that were "actually" created on/after July 1st 2020**, because currently I don't know which attribute I can trust on to be the source of that information.
- Using `UserSettingsModel.created_on` could result in a lot of **false positives** i.e a model that was actually created before July 1st, 2020, but due to earlier flaws has got a timestamp after this cut-off date.
- Rest of the attributes of `UserSettingsModel` can change over time. `first_contribution_msec` remains fixed but it could be None too. And even when it's not None, it has no correlation with creation time.
- Other models were all added over time and some have been added after the cut-off date too, so we can't really trust them for this information.

Considering all of this, I don't see a direct way of identifying them. The **best way according to me** is - If we find that 
  1. `min_date` (among_all_models) <= `user_settings_model.created_on` - `time_delta`, and
  2. `min_date`  >= `cut-off date`

then we **safely assume that the model was created after cut-off date only** (because even the best `min_date` we are updating it to also lies after `cut-off date`) and yield the error that we want (in addition to already being yielded info). I think either we use this or drop the check only, as I see no other way around.

Thoughts?

vojtechjelinek replied:
@sajalasati Ah, good analysis, I agree with your approach, thanks for the great explanation.

sajalasati replied:
@seanlip Kindly resolve this conversation now if you are satisfied with the work done on this PR after you raised your initial concern. Thanks!

seanlip commented:
I have some concerns about this job, so I'm not going to run it yet. Please take a look at the comments above.

Also /cc @vojtechjelinek for info since I want to call attention to the fetch-multi thing.

BenHenning commented:
@seanlip deferring to you for reviewing this since you've already started and I'm handing over codeowners.

@sajalasati feel free to follow up with me directly if my specific input is needed anywhere here.

Commit:
change on-off job

Commit:
resolve merge conflict

sajalasati commented:
@seanlip PTAL now! I have made the changes you had asked for.

sajalasati commented:
@vojtechjelinek @seanlip I need some help in debugging one of the tests (see [here](https://github.com/oppia/oppia/pull/11437/checks?check_run_id=1607819873#step:8:5591)) - where it's saying `AttributeError: 'FixUserSettingsCreatedOnOneOffJobTests' object has no attribute 'assertLessThan'` - can you tell why it's saying so? I can see that this assert statement exists along with ones.

vojtechjelinek commented:
> @vojtechjelinek @seanlip I need some help in debugging one of the tests (see [here](https://github.com/oppia/oppia/pull/11437/checks?check_run_id=1607819873#step:8:5591)) - where it's saying `AttributeError: 'FixUserSettingsCreatedOnOneOffJobTests' object has no attribute 'assertLessThan'` - can you tell why it's saying so? I can see that this assert statement exists along with ones.

Why not use just `assertLess`?

Commit:
fix failing test

Commit:
complete review modifications

Commit:
Merge remote-tracking branch 'upstream/develop' into correct-created-on-attribute

sajalasati commented:
@vojtechjelinek You can do a full review now of all the changes I made!
- I have made some changes in the tests. Also I removed `test_multiple_runs_of_one_off_job_raises_no_error` as it wasn't testing anything in particular. It made sense when we were actually updating the models too, but not now when we only auditing.
- Removed `UserSettingsModel.created_on` from the `user_dates_list`, because we are finding the correct value for this only, and it doesn't make sense to keep it in the list. The list won't become empty ever, as some attributes of UserSettingsModel would be there with values not None.

vojtechjelinek commented on feconf.py:
```diff
@@ -1293,3 +1293,7 @@ def get_empty_ratings():
 # Minimum timestamp difference required between UserSettingsModel's created_on
 # and some other attribute to use it for updating the created_on attribute.
 CREATED_ON_TIME_DELTA_MINUTES = 5
+
+# Date after which UserSettingModel's created_on attribute is supposed to have
+# its correct value.
+CREATED_ON_CORRECTION_CUTOFF_DATE = 'Jul 1 2020'
```
I would place this in the job itself since the job is just temporary and we do not want this constant to be used anywhere else.

sajalasati replied:
Okay then, since I am using it only in the job and in one other test `test_update_for_user_model_created_before_cutoff_date`, I have used the string itself directly.

vojtechjelinek reviewed (changes requested):
LGTM! Just one nit. Keeping Request changes since more code will be added after we test this on the backup server.

Commit:
nit change

Commit:
Merge remote-tracking branch 'upstream/develop' into correct-created-on-attribute

sajalasati commented:
@seanlip PTAL, I have made all the changes like you had mentioned. Should I also go ahead and submit the server jobs form also? Kindly confirm once and I will submit is ASAP.

seanlip commented:
Yup, I'll run it. The old entry is fine. Thanks!

seanlip commented:
@sajalasati, the job is throwing lots of errors. Did you test this manually? Please make sure you run through all the steps of the job on your local machine (with appropriate data set-up beforehand) before requesting a re-run.

```
Traceback (most recent call last): (/third_party/python_libs/mapreduce/handlers.py:517)
  File "/third_party/python_libs/mapreduce/handlers.py", line 503, in handle
    tstate.input_reader, shard_state, tstate, ctx)
  File "/third_party/python_libs/mapreduce/handlers.py", line 593, in _process_inputs
    entity, input_reader, ctx, tstate):
  File "/third_party/python_libs/mapreduce/handlers.py", line 635, in _process_datum
    for output in result:
  File "/core/domain/user_jobs_one_off.py", line 641, in map
    filtered_user_dates_list, key=lambda x: x[1])
TypeError: can't compare datetime.datetime to float
```

sajalasati commented:
@seanlip Ok let me check it again.

sajalasati commented:
@seanlip The `UserSettingsModel.first_contribution_msec` attribute is a `FloatProperty` instead of a `DateTimeProperty` and that's why it is producing errors. I had missed that (and some other attributes of `UserSettingsModel`) in the tests that's why couldn't see I am taking a `FloatProperty` into the list :(

I have fixed the bug now, by converting `Float` timestamp to `DatetimeProperty` using method like [this](https://github.com/oppia/oppia/blob/develop/utils_test.py#L612). All the attributes have been covered now in the tests, and the job also runs fine on local server (with users having both None and not None values of `first_contribution_msec`).

Commit:
fix first_contribution_msec in job

Commit:
Merge remote-tracking branch 'upstream/develop' into correct-created-on-attribute

seanlip commented:
Thanks @sajalasati. I sent you an email with the results.

Commit:
fix merge conflict with RemoveFeedbackThreadIDsOneOffJob

Commit:
job changes completed

Commit:
Merge remote-tracking branch 'upstream/develop' into correct-created-on-attribute

sajalasati commented:
@vojtechjelinek PTAL! I have made the changes in the approach that we discussed offline. In the current approach, I have taken the `time_delta_window` to be 60 minutes, i.e the time window to consider timestamps to be close enough from the absolute minimum timestamp in the list. I hope that value sounds good enough (let me know if it can be increased), because it will help us have a broader set of eligible attributes, so that eventually it also helps us obtain a small set for source of truth.
    The `time_delta_for_update` is left unchanged to be 5 mins. I was not sure whether to increase it or not. Keeping it small would help us discover (and fix) more models which have such inconsistencies. But if a larger value of time gap also sounds good, we would end having to update fewer models.

What do you think about both the values?

sajalasati commented:
@vojtechjelinek As you mentioned in your mail that https://github.com/oppia/oppia/blob/develop/core/domain/base_model_validators.py#L306 can do the check for ensuring that `created_on` can only be upto 1 second greater than `last_updated`, so can I safely assume that via prod validation jobs we know this condition is true for **all models**, and hence drop the `last_updated` attribute for them (except `UserSettingsModel`)? Sorry for taking such a long route in understanding this, I just wanted to make sure that I understand where this check happens. Thanks!

vojtechjelinek commented:
> @vojtechjelinek As you mentioned in your mail that [https://github.com/oppia/oppia/blob/develop/core/domain/base_model_validators.py#L306](https://github.com/oppia/oppia/blob/develop/core/domain/base_model_validators.py?rgh-link-date=2021-01-01T19%3A01%3A16Z#L306) can do the check for ensuring that `created_on` can only be upto 1 second greater than `last_updated`, so can I safely assume that via prod validation jobs we know this condition is true for **all models**, and hence drop the `last_updated` attribute for them (except `UserSettingsModel`)? Sorry for taking such a long route in understanding this, I just wanted to make sure that I understand where this check happens. Thanks!

Yes, the only model where we cannot be sure of is the `UserStatsModel` since we cannot run a prod validation job on it and thus there might be `last_updated` lower than `created_on`.

vojtechjelinek commented on feconf.py:
```diff
@@ -1295,3 +1295,7 @@ def get_empty_ratings():
     SUGGESTION_TYPE_TRANSLATE_CONTENT,
     SUGGESTION_TYPE_ADD_QUESTION
 ]
+
+# Minimum timestamp difference required between UserSettingsModel's created_on
+# and some other attribute to use it for updating the created_on attribute.
+CREATED_ON_TIME_DELTA_MINUTES = 5
```
Why have it here, move it to _user_jobs_one_off.py_

sajalasati replied:
Done now. So basically only the constants which could be used at "many" places in the backend should be kept here right, and not all the constants?

vojtechjelinek reviewed (changes requested):
Thanks! Only one comment for now, you can request another backup server run.

Commit:
nit change

Commit:
Merge remote-tracking branch 'upstream/develop' into correct-created-on-attribute

sajalasati commented:
Hi @vojtechjelinek! Thanks for your review!
Just wanted to double-check that you also read my previous [comment](https://github.com/oppia/oppia/pull/11437#issuecomment-753356084) and you are fine with it.

vojtechjelinek commented:
> Hi @vojtechjelinek! Thanks for your review!
> Just wanted to double-check that you also read my previous [comment](https://github.com/oppia/oppia/pull/11437#issuecomment-753356084) and you are fine with it.

Yes, I'm fine with the values.

sajalasati commented:
Hi @seanlip! Can you please re-run this job on the backup server? I have run this on the local server, and the job runs fine there. Though there, I could only produce the output for the `SUCCESS_ALREADY_UP_TO_DATE` key as it's difficult to check for other keys. In the tests also I have ensured all the datetime fields are populated at least in one of the tests, so I hope you don't see any errors like the previous run. Thanks!

seanlip commented:
@sajalasati Done, I have sent you and @vojtechjelinek the results. Thanks!

sajalasati commented:
> @sajalasati Done, I have sent you and @vojtechjelinek the results. Thanks!

Hi @seanlip ! I have replied to the email with my analysis of the results. @vojtechjelinek agrees with it. So, I will proceed to make code changes for actually updating the models unless you have any objections. Thanks!

seanlip commented:
Yup, happy to defer to @vojtechjelinek on this. Thanks!

Commit:
update job to put() values

Commit:
Merge remote-tracking branch 'upstream/develop' into correct-created-on-attribute

sajalasati commented:
Hi @vojtechjelinek PTAL now! I have update the job based on test results on backup server to also now `put()` the correct values to the datastore now. Updated the PR description accordingly.

vojtechjelinek commented on core/domain/user_jobs_one_off.py:
```diff
@@ -468,6 +469,170 @@ def reduce(key, values):
+        correction_cutoff_timestamp = datetime.datetime.strptime(
+            'Jul 1 2020', '%b %d %Y')
+        if user_settings_model.created_on - min_date > time_delta_for_update:
+            user_settings_model.update_timestamps()
```
```suggestion
            user_settings_model.update_timestamps(update_last_updated_time=False)
```

sajalasati replied:
Changed now. Thanks for the suggestion, I missed an imp thing!

vojtechjelinek commented on core/domain/user_jobs_one_off.py:
```diff
@@ -468,6 +469,170 @@ def reduce(key, values):
+            user_dates_list.append(
+                (
+                    'UserSettingsModel_first_contribution_msec',
+                    datetime.datetime.fromtimestamp(
+                        python_utils.divide(
+                            user_settings_model.first_contribution_msec, 1000.0)
+                    )
```
Don't we have some code for this transformation somewhere already, maybe in utils?

sajalasati replied:
@vojtechjelinek No, I had already checked for it in `utils.py`, but this function is not present. We have similar functions (like converting string to datetime object etc) but not timestamp to datetime object. 
Is it fine this way only or do you want me to first create one function in `utils.py` and then use it here? (We only want to run the job once, but such function could still be useful).

sajalasati replied:
@vojtechjelinek Yes I did check for it `utils.py`, but a function for this specific transformation does not exist as of now there. We have related functions, like converting string to datetime object etc, but not this one.

Is it okay the current way or does it make more sense to create a function in `utils.py` and then use it here (if you think it could be useful elsewhere, because this job only needs to be run once).

vojtechjelinek replied:
Let's keep it here then.

sajalasati replied:
Sure!

vojtechjelinek commented on core/domain/user_jobs_one_off.py:
```diff
@@ -468,6 +469,170 @@ def reduce(key, values):
 
+class FixUserSettingsCreatedOnOneOffJob(jobs.BaseMapReduceOneOffJobManager):
+    """Job that fixes the invalid values of created_on attribute in the
+    UserSettingsModel. It is a one-off job and can be removed from the codebase
+    after we resolve this issue by running the job once.
```
```suggestion
    UserSettingsModel. 
    
    It is a one-off job and can be removed from the codebase after we resolve 
    this issue by running the job once in the January 2021 release.
```

sajalasati replied:
Changed now.

vojtechjelinek commented on core/domain/user_jobs_one_off.py:
```diff
@@ -468,6 +469,170 @@ def reduce(key, values):
+                model_names_and_ids_to_be_fetched_in_batch)
+        )
+
+        for model_name_tuple, model_list in python_utils.ZIP(
+                model_names_and_ids_to_be_fetched_in_batch,
+                fetched_batch_models
+            ):
```
```suggestion
        for model_name_tuple, model_list in python_utils.ZIP(
            model_names_and_ids_to_be_fetched_in_batch,
            fetched_batch_models
        ):
```

sajalasati replied:
After doing your suggestion, I am getting lint error, asking to keep `model_names_and_ids_to_be_fetched_in_batch` and `fetched_batch_models` at 8 spaces from the indentation of `for` (what I had done initially). I have moved the `):` to the previous line - I just had a look [here](https://github.com/oppia/oppia/blob/develop/core/domain/user_jobs_one_off.py#L599-L601) and [here](https://github.com/oppia/oppia/blob/develop/core/domain/user_jobs_one_off.py#L610-L612) and followed the same pattern. I hope that looks good!

vojtechjelinek commented on core/domain/user_jobs_one_off_test.py:
```diff
@@ -1688,6 +1688,280 @@ def test_one_subscription_model_without_feedback_thread_ids(self):
+        self.assertNotEqual(
+            migrated_user_model.created_on, user_auth_details_model.created_on)
+
+    def test_update_for_multiple_users_works_correctly(self):
```
Could you please test a rerun of this job too? So that if it fails for any reason midway we can just run it again and everything works.

sajalasati replied:
Done. Created another test for multiple runs of this, I hope it tests what we wanted correctly. PTAL!

vojtechjelinek reviewed (changes requested):
Thanks! Few last comments.

Commit:
address comments

sajalasati commented:
@vojtechjelinek PTAL! Addressed your comments now.

vojtechjelinek commented:
@sajalasati Thanks! Looks pretty good, one thing I forgot about, we should probably have some kind of audit job for this that can be run before and after to verify that the data are now correct. Probably a job similar to the one that we have now but that will just make sure that `UserSettingsModel.created_on` is lower, with some threshold allowed (`time_delta_for_update`). Probably enough to only check the subset of models that we use now.

After you do that just request another backup server run.

sajalasati commented:
@vojtechjelinek I need to create only one job (say with the name `UserSettingsCreatedOnAuditOneOffJob`), in `user_jobs_one_off.py` only, that uses same set of shortlisted models we arrived at, and ensures the correctness of `created_on`, and this same job will be run before and after running `FixUserSettingsCreatedOnOneOffJob`, right?

Commit:
temp change

vojtechjelinek commented:
@sajalasati Right

Commit:
completed adding new audit job

Commit:
Merge remote-tracking branch 'upstream/develop' into correct-created-on-attribute

sajalasati commented:
@vojtechjelinek I have created the audit job now, PTAL! I have tested both the jobs on local server (for a case of 2 users, with one having correct value, and the other one not - pre-ran audit job, then fix job, and then again audit job, and everything works correctly), and the unit tests also pass on my machine. If the job and the yielded keys looks good to you, I will request another backup server run. Thanks!

@seanlip Should I submit another server jobs form, or simply the editing the instructions in the same google doc attached with the form works?

sajalasati commented:
@seanlip I have updated the job instructions in [google doc](https://docs.google.com/document/d/1mfN7pIMXzKBnXfIZLxE1XZnHylV4knAHOnvIpOwtUmk/edit#). Let me know if anything else is needed!

vojtechjelinek reviewed (approved):
LGTM! @seanlip Please tests this on the backup server and approve if it is okay.

nithusha21 commented:
Not sure why I was assigned... What do you need me to do here?

sajalasati commented:
> Not sure why I was assigned... What do you need me to do here?

Hi @nithusha21! I don't how it shows that sometime back "I requested a review from you as a code owner" (after I merged my branch with develop once). @vojtechjelinek or @seanlip can actually confirm about this. (Though they didn't mention to me earlier about your specific review being needed here, so feel free to skip I maybe)

vojtechjelinek commented:
> Not sure why I was assigned... What do you need me to do here?

@nithusha21  Sorry, I thought you're needed as code owner, unassigned you now.

seanlip commented:
Just a note to say sorry, I'm running a bit late on this. I'll take a look tomorrow!

sajalasati commented:
Hi @seanlip ! Thanks for your note. No issues as such, only that we were hoping that this gets included in the release cut (which is on Friday 8th Jan). But I think we can still make it. Once the job is run, I will be ready to make quick changes, if any needed.

seanlip commented on core/domain/user_jobs_one_off.py:
```diff
@@ -468,6 +469,325 @@ def reduce(key, values):
+        model_name, min_date = min(filtered_user_dates_list, key=lambda x: x[1])
+        time_delta_for_update = datetime.timedelta(minutes=5)
+
+        # This method for converting date_time_string to datettime object has
```
There's a spelling error here: datetime. But fixing it is optional, I think.

sajalasati replied:
@seanlip If I end up doing a follow up PR for [this](https://github.com/oppia/oppia/pull/11437#discussion_r554304442), I will fix this spelling error and the [other](https://github.com/oppia/oppia/pull/11437#discussion_r554283565) change you suggested. Thanks!

seanlip commented on core/domain/user_jobs_one_off.py:
```diff
@@ -468,6 +469,325 @@ def reduce(key, values):
+            (attribute_name, date) for attribute_name, date in user_dates_list
+            if date is not None
+        ]
+        model_name, min_date = min(filtered_user_dates_list, key=lambda x: x[1])
```
Is the list guaranteed to be nonempty? If so, why?

sajalasati replied:
@seanlip I think this is important so I will describe it in a bit detail. 

Here my assumption was that `UserSettingsModel.last_updated` would be present for all models. (I believe `last_agreed_to_terms` and `last_logged_in` should also be at least present, though I was unsure if there exists a scenario where it's possible that one or both of these might not exist). 

With this assumption only I proceeded to exclude `UserSettingsModel.created_on` itself from the list. Though I realise now I could have still included it in the list, without adding any unwanted updates and added complexity overall, and make the overall solution more safe if my assumption isn't always absolutely true.

If you see any problem with these assumptions, I can go ahead and do further analysis, or include `UserSettinhgsModel.created_on` in the list which makes the solution foolproof.

seanlip replied:
Thanks for explaining! I agree with your analysis. I am not too worried though because I didn't see any issues in my test run and all indications suggest that everything was fine for users created within the last 6 months, so I think this is just a theoretical concern.

If you want to be safe, it's fine to do a follow-up PR, but I think it's probably optional. You might want to get @vojtechjelinek's opinion.

seanlip commented on core/domain/user_jobs_one_off.py:
```diff
@@ -468,6 +469,325 @@ def reduce(key, values):
+        if user_settings_model.created_on - min_date > (
+                time_delta_for_correctness):
+            yield (
+                'ERROR_NEED_TO_UPDATE_USING_' + python_utils.UNICODE(
```
Here and elsewhere, always use string interpolation rather than string concatenation (this is a codebase convention).

sajalasati replied:
Sorry, though you had mentioned this to me once before, but I forgot about that.
Does you think it falls under the category of lint checks?

seanlip replied:
Yeah, seems reasonable to add it, I think. Thanks!

sajalasati replied:
Done! Added to Python checks in #8423.

seanlip replied:
Awesome, thanks!

seanlip reviewed (approved):
From a prod server perspective, this looks great! It runs perfectly.

I just have a few comments. They are minor so please consider them optional.

seanlip commented:
@sajalasati Actually, yeah, let's just consider them optional. I'll merge it. Maybe reply to the second one.

Thanks!

sajalasati commented:
@vojtechjelinek I have assigned you because we need your opinion on [this](https://github.com/oppia/oppia/pull/11437#discussion_r554299065) comment - do you feel it's needed to make a follow up PR?
**Note**: If this PR change has already gone into the release cut, and hence if this job is going to be run, we would need to fix this ASAP (which I can do).
Thanks!

vojtechjelinek commented:
@sajalasati I think it is fine without a follow-up PR, I don't remember a case when the `last_updated` field was empty, and the job passed on the backup server so it should be fine.

sajalasati commented:
ok @vojtechjelinek, thanks for confirming!

sajalasati commented:
Unassigned everyone as all the concerns have been addressed.
